### DISCUSSION TITLE
Standardization of filenames and directory structure

### DISCUSSION THREAD

**Discussion by Tapawingo** [clustermod/HCMF3-Loadouts#21] [General]
## Description of Standardization
I think a standardization of filenames would be a good idea to make it more clear. It will also make developing tools for searching and indexing loadouts easier. I suggest the following format:
`[countrycode]_[era]_[unit]_[additionalInfo].sqf`

The entire name should be in lower case for case sensitive searches. This is especially important for Arma instances running on linux as the search is case sensitive.

whitespaces should not be used for the same reason as above.

keywords should be delimited by an underscore `_`.

### countrycode
`Country code` should use the [ISO 3166 Alpha 3](https://www.iban.com/country-codes) country codes to denote the loadout's country. If a country isn't listed in the ISO 3166 standard or if the loadout isn't linked to a particular country other short-format codes should be used. It is important however that these short-format codes do not match any existing country codes from ISO 3166 and thus when one is made it is important that the newly created code is checked against the ISO 3166 list. 
To create some early standardizations for missing codes I suggest the following to be standardized:
- **fic**: Fictional
- **irg**: Irregular
- **pmc**: Private Military Contractor

### era
The `era` would denote a time period for the loadout. It should be one of the following formats (where Y is year):
`YYYY`, `YYYY-YYYY`, `YYs`, `YYs-YYs`, `timeperiod` (e.g. coldwar) or `na` (for Not Applicable)

### unit
`Unit` is an optional parameter, the only requirements are that it is in lowercase. This keyword can be omitted if deemed unnecessary for describing the loadout or if it's not applicable.

### additionalinfo
`Additional Info` is an optional parameter, the only requirements are that it is in lowercase. This keyword can be omitted if deemed unnecessary for describing the loadout or if it's not applicable.

**Comment by Tapawingo** [comment]
As an addition to this standardization I also suggest a new directory structure to make it easier for the standard user to locate loadout they want:
```
loadouts
├── [Country] (e.g. germany)
│   ├── loadout (contains all player loadouts)
│   │   ├── [Kosher Arsenal Loadout] (template file)
│   │   └── README.md (overview of loadouts, and relevant garage, logistics or ai templates)
│   ├── garage
│   │   ├── [Kosher Garage Loadout] (template file)
│   │   └── README.md (overview of loadouts, and relevant garage, logistics or ai templates)
│   ├── logistics
│   │   ├── [Logistics Template] (template file)
│   │   └── README.md (overview of loadouts, and relevant garage, logistics or ai templates)
│   └── ai
│       ├── [Kosher AI Loadout] (template file)
│       └── README.md (overview of loadouts, and relevant garage, logistics or ai templates)
└── README.md (for documentation of this directory)
```

The README in the `loadouts` directory is mainly for creating a route selection for users. Idea being they can navigate through the repository using links in the readme files. This way you can put up a lot more information to make it easier for someone unfamiliar with GitHub to find what they are looking for. Additionally this can be used for documentation of the directory structure or file structure.

The README inside the `sub-directories` of the `Country` directories would serve to list each of the files, information about the files, relevant or associated templates (garage, logistics or AI) and a direct download link of the files. This again would work into the navigation system described above making it easier for the standard user to more easily find what they are searching for.

**Reply by PervonHarke** [reply]
I think having a easily readable list that directly downloads that file when clicked on,as we spoke about in Discord, is the best solution. 

Maybe the readme file would be a good place to put that. Below all the other stuff that is already written there.

**Reply by PervonHarke** [reply]
A possible addition to this.

I think it would be better if we would do the mission sets not with folders but with a Readme and those direct download links. Because else one would need to constantly update two files if they change them. Maybe not that major but I could see me maybe forgetting to update one.

**Comment by Tapawingo** [comment]
These rules can be enforced by creating [contributor guidelines](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors) and amending the [Pull Request template](https://github.com/clustermod/Harks-CMF3-Loadouts/blob/master/.github/PULL_REQUEST_TEMPLATE.md) to enforce these rules.

Additionally, a tool to verify filenames and directory structure could very easily be made that a contributor can run in vscode and that could also be run on new pull requests to verify them before merging

**Comment by Tapawingo** [comment]
Added a branch where you can see an example of the changes I'm suggesting. filenames and structure has been fully completed. as for the READMEs only the path leading down to Norwegian loadouts have been completed. You can preview it here: [standardizationExample branch](https://github.com/clustermod/Harks-CMF3-Loadouts/tree/standardizationExample)

**Comment by PervonHarke** [comment]
## Amendment 

Adding _camouflage_ to the loadout file name as a mandatory parameter.
`[countrycode]_[era]_[unit]__[camouflage]_[additionalInfo].sqf`

### Camouflage
is a mandatory parameter that describes the most prevalent camouflage of the loadout. The standardized codes are:
- **mixed:** used if there is no major camo pattern in every kit in the loadout file.
- **civ:** used if the loadout only has civilian clothing in it with minor appearances of camouflaged items.

Other camouflages are to be written in length (ie.: Flecktarn, Multicam). If the names are too long or have spaces in their name they are to be shortened to the most commonly used acronym for them _(ie.: Desert Nigth Camouflage --> DNC)_. 

***
For the file structure, I had the idea of having packs for each unit containing the three files (loadout, logistics, garage) to be very quickly downloaded and thrown into a mission to avoid mixups and decrease the chance of someone putting in the wrong files.

But I see an issue with edge cases where there is no garage and logistics file which would create an ugly file structure.
So I think it is a good decision to change the current one to your proposed new structure.

**Comment by PervonHarke** [comment]
I have added slightly changed versions of the following categories, copied from your messages above @Tapawingo  to the wiki of this repository. They are now set standards. 
- [Description of Standardization](https://github.com/clustermod/Harks-CMF3-Loadouts/wiki#standardization)
- [Format and Syntax Description](https://github.com/clustermod/Harks-CMF3-Loadouts/wiki#format-and-syntax-description)
- [File Structure](https://github.com/clustermod/Harks-CMF3-Loadouts/wiki#format-and-syntax-description)

**Comment by PervonHarke** [comment]
I just noticed that the version number for the Cluster Community Modpack is not in the standardized naming format. 
This saves a lot of times for mission makers not needing to manually test if the loadout is missing items in the current CC modpack. 
Especially when they are not familiar with that would be koscher for the depicted unit. The only option would be to manually go through every item and check if it is present in the mission/modpack.

**Reply by Tapawingo** [reply]
well I thought I'd make the tool you requested some months ago in CMF to create a list of dependent mods. And then we can use that tool to see if the loadouts are up to date with our current modpack at least. But I'm not sure when I'll get around to it.

**Reply by PervonHarke** [reply]
I will implement a standardized file header in the loadout/garage/logistic sqf files. As a workaround for now. 
 Example of said header:
```
/*
 * Made for Arma 3 Virtual MILSIM
 *
 * Author: Hark
 * CC Modpack Version: V5
 * [Description/Extra Information]
 *  
 *
 */
```
